pyp2rpm/extract_distribution.py

```diff
@@ -0,0 +1,76 @@
+"""
+Modify the Distutils 'bdist_rpm' command (create RPM source and binary
+distributions) to be able of automatic rebuild in Fedora's COPR and
+fit better to Fedora packaging Guidlines.
+"""
+
+import distutils.command.bdist_rpm
+import glob
+import re
+import time
+import sys
+import os.path
+import runpy
+
+bdist_rpm_orig = distutils.command.bdist_rpm.bdist_rpm
+
+
+class extract_distribution(bdist_rpm_orig):
+    class_distribution = None
+
+    def finalize_package_data (self):
+        self.distribution.force_arch = self.force_arch
+
+        self.distribution.build_requires = self.build_requires or (
+                               self._list(getattr(self.distribution, 'setup_requires', []))
+                               + self._list(getattr(self.distribution, 'tests_require', [])))
+
+        self.distribution.run_requires = self.requires or self._list(getattr(
+            self.distribution, 'install_requires', []))
+
+        self.distribution.conflicts = [dep.replace('!=', '=') 
+                                      for dep in self.distribution.run_requires if '!=' in dep]
+
+        if (getattr(self.distribution, 'entry_points', None)
+            and 'setuptools' not in self.distribution.run_requires):
+            self.distribution.run_requires.append('setuptools')
+
+        self.distribution.distribution_name = self.distribution
+
+        script_options = [ 
+            ('prep', 'prep_script'),
+            ('build', 'build_script'),
+            ('install', 'install_script'),
+            ('clean', 'clean_script'),
+        ]
+
+        for (rpm_opt, attr) in script_options:
+            val = getattr(self, attr, None)
+            if val:
+                setattr(self.distribution, rpm_opt, val) 
+
+        if self.distribution.metadata.classifiers is None:
+            self.distribution.metadata.classifiers = []
+
+    def run(self):
+        extract_distribution.class_distribution = self.distribution
+
+    @staticmethod
+    def _list(var):
+        if var is None:
+            return []
+        elif not isinstance(var, list):
+            raise DistutilsOptionError("{} is not a list".format(var))
+        return var 
+
+distutils.command.bdist_rpm.bdist_rpm = extract_distribution
```

Comment:
extracted setup.py is run like this in DistMetadataExtractor's __init__ method:
`extract_distribution.run_setup(setup_py, 'bdist_rpm')`

extract_distribution command is executed instead of bdist_rpm thanks to this  assignment.
There is other way to do it using --command-packages option:
https://docs.python.org/2/distutils/extending.html
,but this doesn't work for me, I am not sure why.